app/services/katello/pulp/pulp_content_unit.rb

```diff
@@ -22,10 +22,143 @@ def self.pulp_data(uuid)
+
+      def self.import_for_repository(repository)
+        pulp_ids = []
+        pulp_units_batch_for_repo(repository).each do |units|
+          units.each do |unit|
+            unit = unit.with_indifferent_access
+            model = Katello::Util::Support.active_record_retry do
+              model_class.where(:pulp_id => unit['_id']).first_or_create
+            end
+            service = self.new(model.pulp_id)
+            service.backend_data = unit
+            service.update_model(model)
+            pulp_ids << model.pulp_id
+          end
+        end
+        sync_repository_associations(repository, :pulp_ids => pulp_ids) if model_class.manage_repository_association
+      end
+
+      def self.sync_repository_associations(repository, options = {})
+        additive = options.fetch(:additive, false)
+        associated_ids = options.fetch(:ids, nil)
+        pulp_ids = options.fetch(:pulp_ids) if associated_ids.nil?
+        associated_ids = model_class.with_pulp_id(pulp_ids).pluck(:id) if pulp_ids
+        table_name = model_class.repository_association_class.table_name
+        attribute_name = model_class.unit_id_field
+
+        existing_ids = model_class.repository_association_class.uncached do
+          model_class.repository_association_class.where(:repository_id => repository).pluck(attribute_name)
+        end
+        new_ids = associated_ids - existing_ids
+        delete_ids = existing_ids - associated_ids
+
+        queries = []
+
+        if delete_ids.any? && !additive
+          queries << "DELETE FROM #{table_name} WHERE repository_id=#{repository.id} AND #{attribute_name} IN (#{delete_ids.join(', ')})"
+        end
+
+        unless new_ids.empty?
+          inserts = new_ids.map { |unit_id| "(#{unit_id.to_i}, #{repository.id.to_i}, '#{Time.now.utc.to_s(:db)}', '#{Time.now.utc.to_s(:db)}')" }
+          queries << "INSERT INTO #{table_name} (#{attribute_name}, repository_id, created_at, updated_at) VALUES #{inserts.join(', ')}"
+        end
+
+        ActiveRecord::Base.transaction do
+          queries.each do |query|
+            ActiveRecord::Base.connection.execute(query)
+          end
+        end
+      end
+
+      def self.update_repository_associations(units_json, additive = false)
+        ActiveRecord::Base.transaction do
+          repo_unit_id = {}
+          repo_cache = {}
+          units_json.each do |unit_json|
+            unit_json['repository_memberships'].each do |repo_pulp_id|
+              if (repo_cache[repo_pulp_id] ||= Katello::Repository.exists?(:pulp_id => repo_pulp_id))
+                repo_unit_id[repo_pulp_id] ||= []
+                repo_unit_id[repo_pulp_id] << unit_json['_id']
+              end
+            end
+          end
+
+          repo_unit_id.each do |repo_pulp_id, unit_pulp_ids|
+            sync_repository_associations(Katello::Repository.find_by(:pulp_id => repo_pulp_id), :pulp_ids => unit_pulp_ids, :additive => additive)
+          end
+        end
+      end
```

Comment:
we're only supporting indexing a content unit and associating it with a single repository, so this can probably be simplified?

Reply:
The additive option seemed unnecessary after all the changes. Now we are getting ids to add and ids to remove, forming queries and executing them.